src/reference/00-Getting-Started/06-Task-Graph.md

```diff
@@ -0,0 +1,427 @@
+
+The basic Makefile syntax looks like the following:
+
+```
+target: dependencies
+[tab] system command1
+[tab] system command2
+```
+
+Given a target (the default target is named `all`),
+
+1. Make checks if the target's dependencies have been built, and builds any of the dependencies that hasn't been built yet.
+2. Make runs the system commands in order.
+
+Let's take a look at a `Makefile`:
+
+```
+CC=g++
+CFLAGS=-Wall
+
+all: hello
+
+hello: main.o hello.o
+    \$(CC) main.o hello.o -o hello
+
+%.o: %.cpp
+    \$(CC) \$(CFLAGS) -c \$< -o \$@
+```
+
+Running `make`, it will by default pick the target named `all`.
+The target lists `hello` as its dependency, which hasn't been built yet, so Make will build `hello`.
+
+Next, Make checks if the `hello` target's dependencies have been built yet.
+`hello` lists two targets: `main.o` and `hello.o`.
+Once those targets are created using the last pattern matching rule,
+only then the system command is executed to link `main.o` and `hello.o` to `hello`.
+
+If you're just running `make`, you can focus on what you want as the target,
+and the exact timing and commands necessary to build the intermediate products are figured out by Make.
+We can think of this as dependency-oriented programming, or flow-based programming.
+Make is actually considered a hybrid system because while the DSL describes the task dependencies, the actions are delegated to system commands.
+
+#### Rake
+
+This hybridity is continued for Make successors such as Ant, Rake, and sbt.
+Take a look at the basic syntax for Rakefile:
+
+```ruby
+task name: [:prereq1, :prereq2] do |t|
+  # actions (may reference prereq as t.name etc)
+end
+```
+
+The breakthrough made with Rake was that it used a programming language to
+describe the actions instead of the system commands.
+
+#### Benefits of hybrid flow-based programming
+
+There are several motivation to organizing the build this way.
+
+First is de-duplication. With flow-based programming, a task is executed only once even when it is depended by multiple tasks.
+For example, even when multiple tasks along the task graph depend on `compile in Compile`,
+the compilation will be executed exactly once.
+
+Second is parallel processing. Using the task graph, the task engine can
+schedule mutually non-dependent tasks in parallel.
+
+Third is the separation of concern and the flexibility.
+The task graph lets the build user wire the tasks together in different ways,
+while sbt and plugins can provide various features such as compilation and
+library dependency management as functions that can be reused.
+
+### Summary
+
+The core data structure of the build definition is a DAG of tasks,
+where the edges denote happens-before relationships.
+`build.sbt` is a DSL designed to express dependency-oriented programming,
+or flow-based programming, similar to `Makefile` and `Rakefile`.
+
+The key motivation for the flow-based programming is de-duplication,
+parallel processing, and customizability.
```

Comment:
Maybe it would be interesting to also add a section explaining why it is possible to write:

~~~ scala
foo := bar.value
~~~

But not:

~~~ scala
val foo = bar.value
~~~

Though the following is possible:

~~~ scala
val foo = Def.setting(bar.value)
~~~

In other words: what’s the magic happening in `:=` that makes it possible to omit the surrounding `Def.setting`. Maybe that could help understanding the `Def.task(…).taskValue` trick, too…